google/cloud/spanner_v1/_helpers.py

```diff
@@ -234,18 +249,38 @@ def _parse_value_pb(value_pb, field_type):
         return DatetimeWithNanoseconds.from_rfc3339(value_pb.string_value)
     elif type_code == TypeCode.ARRAY:
         return [
-            _parse_value_pb(item_pb, field_type.array_element_type)
+            _parse_value_pb(
+                item_pb, field_type.array_element_type, field_name, column_info
+            )
             for item_pb in value_pb.list_value.values
         ]
     elif type_code == TypeCode.STRUCT:
         return [
-            _parse_value_pb(item_pb, field_type.struct_type.fields[i].type_)
+            _parse_value_pb(
+                item_pb, field_type.struct_type.fields[i].type_, field_name, column_info
+            )
             for (i, item_pb) in enumerate(value_pb.list_value.values)
         ]
     elif type_code == TypeCode.NUMERIC:
         return decimal.Decimal(value_pb.string_value)
     elif type_code == TypeCode.JSON:
         return JsonObject.from_str(value_pb.string_value)
+    elif type_code == TypeCode.PROTO:
+        bytes_value = base64.b64decode(value_pb.string_value)
+        if column_info is not None and column_info.get(field_name) is not None:
+            proto_message = column_info.get(field_name)
+            if isinstance(proto_message, Message):
+                proto_message = proto_message.__deepcopy__()
```

Comment:
Why are we doing `__deepcopy__ ` here? Usually the user will be passing the default object which won't be having any value associated with it? 
- Does `ParseFromString ` method completely ignores the value which are present internally and create a new object out of `bytes` or merge in the existing object?

Reply:
```ParseFromString clears the object first and then merges the data into the object. It does not create a new object/instance.```
Then you should use the shallow copy, if you are clearing the values then I don't see any value in deep copying the object.

Reply:
Shallow copies of proto messages are not possible. A shallow copy means constructing a new object and then populating it with references to the child objects found in the original. Let's say you have a message with in a message and now if you shallow copy you get the new object of the outer message, but for the inner message you share the reference to that in the original. This has so many complications and is not used for proto messages.

This is why deep copy is preferred and used.

Reply:
Shallow copy of Proto messages are not possible. A shallow copy of an object is a copy whose properties share the same references as those of the source object from which the copy was made. This implies, if a proto message is shallow copied from another, then both of it would be pointing to same inner references. Hence going forward with deep copy.